diff -Nur Mail-SpamAssassin-3.0.2.orig/lib/Mail/SpamAssassin/HTML.pm Mail-SpamAssassin-3.0.2/lib/Mail/SpamAssassin/HTML.pm
--- Mail-SpamAssassin-3.0.2.orig/lib/Mail/SpamAssassin/HTML.pm	2004-10-22 17:21:02.000000000 +0200
+++ Mail-SpamAssassin-3.0.2/lib/Mail/SpamAssassin/HTML.pm	2004-12-21 18:02:02.000000000 +0100
@@ -176,6 +176,7 @@
 		],
 		marked_sections => 1);

+  $hp->utf8_mode(1);
   # ALWAYS pack it into byte-representation, even if we're using 'use bytes',
   # since the HTML::Parser object may use Unicode internally.
   # (bug 1417, maybe)
