--- spamd/pld-rc-script.sh.old	Thu Sep  5 14:19:19 2002
+++ spamd/pld-rc-script.sh	Thu Sep  5 14:25:26 2002
@@ -4,11 +4,11 @@
 #
 # chkconfig: 2345 80 30
 #
-# description: spamd is a daemon process which uses SpamAssassin to check
-#              email messages for SPAM.  It is normally called by spamc
-#	           from a MDA.
-# processname:	spamassassin
-# pidfile:	/var/run/spamassassin.pid
+# description: spamd is a daemon process which uses SpamAssassin to check \
+#              email messages for SPAM.  It is normally called by spamc \
+#              from a MDA.
+# processname: spamassassin
+# pidfile:     /var/run/spamassassin.pid
 
 # Source function library.
 . /etc/rc.d/init.d/functions
@@ -43,12 +43,12 @@
   stop)
 	# Stop daemons.
 	if [ -f /var/lock/subsys/spamd ]; then
-		echo -n "Shutting down spamd: "
+		msg_stopping SpamAssassin
 		killproc spamd
 		RETVAL=$?
 		rm -f /var/lock/subsys/spamd
 	else
-		msg_Already_Running SpamAssassin
+		msg_Not_Running SpamAssassin
 	fi
 	;;
   restart)
